src/tools/pbs_upgrade_job.c

```diff
@@ -234,102 +311,284 @@ check_job_file(int fd)
+	statechar = state_int2char(qs_19_20.ji_state);
+	if (statechar != JOB_STATE_LTR_UNKNOWN) {
+		bool stateset = false;
+		bool substateset = false;
+
+		snprintf(statebuf, sizeof(statebuf), "%c", statechar);
+		snprintf(ssbuf, sizeof(ssbuf), "%d", qs_19_20.ji_substate);
+		pal = read_all_attrs_from_jbfile(fd, NULL, NULL);
+		pali = pal;
+		while (pali != NULL) {
+			if (strcmp(pali->al_name, ATTR_state) == 0) {
+				pali->al_valln = strlen(statebuf) + 1;
+				pali->al_value = pali->al_name + pali->al_nameln + pali->al_rescln;
+				strcpy(pali->al_value, statebuf);
+				pali->al_tsize = sizeof(svrattrl) + pali->al_nameln + pali->al_valln;
+				stateset = true;
+				if (substateset)
+					break;
+			}
+			else if (strcmp(pali->al_name, ATTR_substate) == 0) {
+				pali->al_valln = strlen(ssbuf) + 1;
+				pali->al_value = pali->al_name + pali->al_nameln + pali->al_rescln;
+				strcpy(pali->al_value, ssbuf);
+				pali->al_tsize = sizeof(svrattrl) + pali->al_nameln + pali->al_valln;
+				substateset = true;
+				if (stateset)
+					break;
+			}
+			if (pali->al_link.ll_next == NULL)
+				break;
+			pali = GET_NEXT(pali->al_link);
+		}
+	}
 
 	/* Open a temporary file to stage data */
 	tmp = tmpfile();
 	if (!tmp) {
-		fprintf(stderr, "Failed to open temporary file [%s]\n",
-				errno ? strerror(errno) : "No error");
+		fprintf(stderr, "Failed to open temporary file [%s]\n", errno ? strerror(errno) : "No error");
 		return 1;
 	}
 	tmpfd = fileno(tmp);
 	if (tmpfd < 0) {
-		fprintf(stderr, "Failed to find temporary file descriptor [%s]\n",
-				errno ? strerror(errno) : "No error");
+		fprintf(stderr, "Failed to find temporary file descriptor [%s]\n", errno ? strerror(errno) : "No error");
 		return 1;
 	}
 
 	/* Write the new jobfix structure to the output file */
 	len = write(tmpfd, &new_job.ji_qs, sizeof(new_job.ji_qs));
 	if (len != sizeof(new_job.ji_qs)) {
-		fprintf(stderr, "Failed to write jobfix to output file [%s]\n",
+		fprintf(stderr, "Failed to write jobfix to output file [%s]\n", errno ? strerror(errno) : "No error");
+		return 1;
+	}
+
+	/* Write the new extend structure to the output file */
+	len = write(tmpfd, &new_job.ji_extended, sizeof(new_job.ji_extended));
+	if (len != sizeof(new_job.ji_extended)) {
+		fprintf(stderr, "Failed to write job extend data to output file [%s]\n",
 				errno ? strerror(errno) : "No error");
 		return 1;
 	}
 
+	/* Write the job attribute list to the output file */
+	pali = pal;
+	while (pali != NULL) {	/* Modeled after save_struct() */
+		int copysize;
+		int objsize;
+
+		objsize = pali->al_tsize;
+		charstrm = (char *) pali;
+		while (objsize > 0) {
+			if (objsize > BUFSZ)
+				copysize = BUFSZ;
+			else
+				copysize = objsize;
+			memcpy(buf, charstrm, copysize);
+			len = write(tmpfd, buf, copysize);
+			if (len < 0) {
+				fprintf(stderr, "Failed to write output file [%s]\n", errno ? strerror(errno) : "No error");
+				return 1;
+			}
+			objsize -= copysize;
```

Comment:
I'd use len  returned by write instead of copysize because theoretically, write can write fewer bytes than specified.